### PR TITLE
ARM Base Image Update Pipeline

### PR DESCRIPTION
**Pull request by yophilav**
First take on automating ARM Base Image update.

Use case scenario: 
1. tools/BaseImageUpdate/baseImage.ps1 is a helper script that a developer can use on his powershell to update the ARM base image version.  i.e. `Update-ARM-BaseImages -NewASPNetCoreVersion "2.1.23"`.
2. A developer send out a Pull request with base image changes.
3. A developer uses a pipeline `builds/misc/base-images-publish.yaml` to kick off the build & base image publication.
4. (Optional) As a good measure, once (3) is completed, a developer can run the build image pipeline to verify a successful publication.

### CODE REVIEW

File: builds/misc/templates/base-image-publish/base-image-build-and-publish.yaml
Comment (yophilav) [inline builds/misc/templates/base-image-publish/base-image-build-and-publish.yaml:29]: I intentionally leave this here for the moment as I'm expecting some comment & code change from this PR. I don't want to accidentally release a base image while testing the code change.

File: builds/misc/templates/base-image-publish/base-image-config.yaml
Comment (huguesBouvier) [inline builds/misc/templates/base-image-publish/base-image-config.yaml:9]: shouldn't the API proxy be in that list?

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [inline builds/misc/templates/base-image-publish/base-image-overview.yaml:14]: I think we can disambiguate what Item3  and this value is by creating a variable for it with a label before this line. We could use this approach in other places as well.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [inline builds/misc/templates/base-image-publish/base-image-overview.yaml:15]: It is possible for this to get written to console, but this should fail the build if it evaluates to false right?

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [inline builds/misc/templates/base-image-publish/base-image-overview.yaml:16]: Why do we set the var to itself? Or am I missing something?

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [inline builds/misc/templates/base-image-publish/base-image-overview.yaml:25]: Is that a typo? Why just arm64?

File: builds/misc/base-images-publish.yaml
Comment (and-rewsmith) [inline builds/misc/base-images-publish.yaml:24]: Why do we need to clean this dir?

File: builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml
Comment (and-rewsmith) [inline builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml:4]: We will need to remember to update this file whenever we increase the rocksdb version

File: builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml
Comment (and-rewsmith) [inline builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml:15]: Why do we need to check this explicitly? Wouldn't the copy step fail otherwise?

File: builds/misc/templates/base-image-publish/base-image-config.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-config.yaml:9]: API proxy uses `arm32v7/haproxy:1.8.23` for ARM32 and `haproxy:1.8.23-alpine ` for AMD64. 
I don't think they are relevant to ARM base images `3.1.10-bionic-arm*` that we are updating here.

With that said, I have a few follow up question: 
1. How do we known when we should upgrade the base image for HAProxy e.g. OS security patch, bug fixes?
2. Are the HAProxy base image somehow related to the Bionic arm image releases? Are they release about the same frequency?

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [inline builds/misc/templates/base-image-publish/base-image-overview.yaml:14]: Why do we set item3 here? Where do we use it later? It seems from the build script all we care about is the var name based on this with the "Version" suffix.

**Review by and-rewsmith — COMMENTED** [review]
Nice work! This will really help us all out. I think step 4 you listed in the description is a great idea, and we should always do it before merging PRs which use this new process.

Has this been ran to completion yet? What changes are you planning on making in following PRs?

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml:8]: `$eachInfo` ? lol.

File: builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml:4]: Yup. 
TODO: Create a documentation about this tool (and address this concern).

File: builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-setup-librocksdb.yaml:15]: It may or may not (depending on the return code). I'd rather have it double check here and print out to console.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml:8]: I think something that indicates that this info is specific to one base image would help people unfamiliar with the whole process. Maybe baseImageContext?

File: builds/misc/base-images-publish.yaml
Comment (yophilav) [reply builds/misc/base-images-publish.yaml]: Removed the file.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: Changed to `config`
For more explicit & convention on the procedure, I'll defer to a user manual (which is to be written).

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: That would be a lot of duplication say... explicitly assigning {`edgeAgentTag`, `edgeHubTag`, `moduleTag`, `fullModuleTag`}. 

Note these variable are only set as env vars for this instance of `pwsh`, they are gone after this task is done.

Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: Per the comment above, this is setting a variable name (value of item3) as a local var to be used in the next line 2 lines of code.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: The wonder of VSTS variable. This is actually setting the variable to pipeline's JOB scope. Which is different that task env var. As it turns out, we don't need this var, so I'm deleting it.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: Refactoring plunder! (a typo)

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: > That would be a lot of duplication say... explicitly assigning {edgeAgentTag, edgeHubTag, moduleTag, fullModuleTag}.

Yeah this is what I had in mind actually. It would have definitely helped me understand what this was doing better.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: I see the output is:
> azureiotedge-agent-base:1.0.6.4-linux-arm32v7 <= 3.1.10-bionic-arm32v7

Is this less than or equals sign supposed to be an arrow?

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: nit: Still think we could make it more clear that this config is specific to one base image. The user manual makes sense, but I think the code can still be more self-documenting.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: Nope I'm not unrolling a loop with 5/6 lines repeating the same code over and over. 
What you request will turn this loop into this: 
```
      $pathPrefix = "$(Build.SourcesDirectory)/edge-agent/docker/$(os)/$(arch)"
      $unberlyingFilePath = $pathPrefix + '/base/Dockerfile' | Resolve-path
      $underlyingTag = $($(Get-Content $unberlyingFilePath | Select-String "ARG base_tag=") -split "=")[1]
      $imageFilePath = $pathPrefix + '/Dockerfile' | Resolve-path
      Set-Variable -Name "edgeAgentTag" -Value "$($($(Get-Content $imageFilePath | Select-String "ARG base_tag=") -split "=")[1])"
      Write-Output "azureiotedge-agent-base:$(Get-Variable -Name "edgeAgentTag" -ValueOnly) 🠘 $underlyingTag"


      $pathPrefix = "$(Build.SourcesDirectory)/edge-hub/docker/$(os)/$(arch)"
      $unberlyingFilePath = $pathPrefix + '/base/Dockerfile' | Resolve-path
      $underlyingTag = $($(Get-Content $unberlyingFilePath | Select-String "ARG base_tag=") -split "=")[1]
      $imageFilePath = $pathPrefix + '/Dockerfile' | Resolve-path
      Set-Variable -Name "edgeHubTag" -Value "$($($(Get-Content $imageFilePath | Select-String "ARG base_tag=") -split "=")[1])"
      Write-Output "azureiotedge-hub-base:$(Get-Variable -Name "edgeHubTag" -ValueOnly) 🠘 $underlyingTag"


      $pathPrefix = "$(Build.SourcesDirectory)/edge-modules/SimulatedTemperatureSensor/docker/$(os)/$(arch)"
      $unberlyingFilePath = $pathPrefix + '/base/Dockerfile' | Resolve-path
      $underlyingTag = $($(Get-Content $unberlyingFilePath | Select-String "ARG base_tag=") -split "=")[1]
      $imageFilePath = $pathPrefix + '/Dockerfile' | Resolve-path
      Set-Variable -Name "moduleTag" -Value "$($($(Get-Content $imageFilePath | Select-String "ARG base_tag=") -split "=")[1])"
      Write-Output "azureiotedge-module-base:$(Get-Variable -Name "moduleTag" -ValueOnly) 🠘 $underlyingTag"


      $pathPrefix = "$(Build.SourcesDirectory)/test/modules/TestAnalyzer/SimulatedTemperatureSensor/docker/$(os)/$(arch)"
      $unberlyingFilePath = $pathPrefix + '/base/Dockerfile' | Resolve-path
      $underlyingTag = $($(Get-Content $unberlyingFilePath | Select-String "ARG base_tag=") -split "=")[1]
      $imageFilePath = $pathPrefix + '/Dockerfile' | Resolve-path
      Set-Variable -Name "fullModuleTag" -Value "$($($(Get-Content $imageFilePath | Select-String "ARG base_tag=") -split "=")[1])"
      Write-Output "azureiotedge-module-base-full:$(Get-Variable -Name "fullModuleTag" -ValueOnly) 🠘 $underlyingTag"
```
I'd rather have a list somewhere and formulate the necessary variable from the list; rather than have multiple duplication of code everywhere you need it.

Comment (and-rewsmith) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: Oh I didn't mean unrolling, I just meant adding labels to make it more readable.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: I see where you're coming from. Without a paragraph explaining that each "base image" has an underlying OS base image associate with them. I'm not sure how else I can explain it. The user should read the code to see what it is doing. Otherwise, you will run into the maintenance hassle between (code comment vs code vs document). I drew the line that anything conceptual & user manual are going to be explained in document. For more detailed implementation, let the user read the code. Naming convention can help so much when a user doesn't have an understand of what we are trying to do.

File: builds/misc/templates/base-image-publish/base-image-overview.yaml
Comment (and-rewsmith) [reply builds/misc/templates/base-image-publish/base-image-overview.yaml]: For example these lines took me a few minutes to understand:
```
      Set-Variable -Name "$($eachConfig.Item3.ToString())" -Value "$($($(Get-Content $imageFilePath | Select-String "ARG base_tag=") -split "=")[1])"
      Write-Output "$($eachConfig.Item2.ToString()):$(Get-Variable -Name "$($eachConfig.Item3.ToString())" -ValueOnly) 🠘 $underlyingTag"
```

I think having labels above these two lines for the values you are using within them would help a lot.

**Review by and-rewsmith — APPROVED** [review]
LGTM. Someone more familiar with devops might have some insight on how to clean it up further.

File: tools/BaseImageUpdate/baseImage.ps1
Comment (damonbarry) [inline tools/BaseImageUpdate/baseImage.ps1:67]: I worry about encoding problems here. I wonder if powershell might insert a BOM in some cases, or change the encoding to utf16 or something. Should we `Get-Content` and `Set-Content` with the `-Encoding UTF8` argument?

**Review by damonbarry — CHANGES_REQUESTED** [review]
In general, I wonder why the pipeline is broken into so many little templates? There's _a lot_ of extra complexity in here to serialize/deserialize and pass values between templates using the `##vso...` syntax, and to check for expected input values at every step along the way. I think the code would be way simpler and easier to read if `base-image-config`, `base-image-overview`, `base-image-*-librocksdb`, and `base-image-build-and-publish` were all one script.

Templates are for reusing tasks in different pipelines, but these tasks aren't shared with any other pipeline.

File: tools/BaseImageUpdate/baseImage.ps1
Comment (yophilav) [reply tools/BaseImageUpdate/baseImage.ps1:15]: Yup. It seems like `$env:BUILD_SOURCESDIRECTORY` is not defined neither $PSScriptRoot

File: builds/misc/templates/base-image-publish/base-image-docker-login.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-docker-login.yaml:9]: This is a tricky one. I have tested the login with the Dokcer@2 task but I couldn't figure out how to get the `connection` to go properly given that we are not logging into Azure Container Registry. We are logging into `https://hub.docker.com/r/azureiotedge/` (aka docker.io)

File: tools/BaseImageUpdate/baseImage.ps1
Comment (damonbarry) [reply tools/BaseImageUpdate/baseImage.ps1:15]: `$env:BUILD_SOURCESDIRECTORY` is only defined in Azure pipelines builds. `$PSScriptRoot` is an automatic variable that contains the directory from which a script is being run, since PowerShell 3.0. We're already using this logic in other scripts.

Comment (yophilav) [reply tools/BaseImageUpdate/baseImage.ps1:15]: > $PSScriptRoot is an automatic variable that contains the directory from which a script is being run

Ah I see. That's why it's returning empty when I tried running it striaght on the pwsh.

File: builds/misc/templates/base-image-publish/base-image-docker-login.yaml
Comment (damonbarry) [reply builds/misc/templates/base-image-publish/base-image-docker-login.yaml:9]: You have to create a new Docker service connection in Azure pipelines. In the example I gave earlier, someone created a service connection called "iotedge-edgebuilds-acr" that connects to edgebuilds. You'd have to do the same, only for docker.io. Something like `containerRegistry: docker.io` won't work because there's no service connection called "docker.io" in our Azure pipelines...you have to create one.

File: builds/misc/templates/base-image-publish/base-image-setup.yaml
Comment (damonbarry) [inline builds/misc/templates/base-image-publish/base-image-setup.yaml:1]: This is better, but you're still distributing the work across multiple templates so you're still doing all the serialization/deserialization. Templates are good for sharing code, but they're bad for organizing code because passing information between templates requires a lot of effort. And you're not sharing this logic with any other pipelines so there's not a compelling reason to use templates. If you want organization, you can use powershell functions inside a single script.

`base-image-build-and-publish.yaml`, for example, is only about 7 lines of code if you remove the deserialization and input checking logic. You could easily merge the powershell scripts from `base-image-setup.yaml`, `base-image-build-and-publish.yaml`, and `base-image-remove-librocksdb.yaml` into one script, and it would make things a lot easier to read and maintain.

And since

File: tools/BaseImageUpdate/baseImage.ps1
Comment (damonbarry) [reply tools/BaseImageUpdate/baseImage.ps1]: Oh right, it's only available inside a script :)

File: builds/misc/templates/base-image-publish/base-image-setup.yaml
Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-setup.yaml:1]: 😆 That how it was all started by one giant yaml script. Looks like this pipeline has come a full circle: dev -> create one giant yaml script -> thinking the script is too large -> overly refactored scripts -> decided the first version is the most readible one -> create one giant script.

Comment (damonbarry) [reply builds/misc/templates/base-image-publish/base-image-setup.yaml:1]: Haha...you definitely need organization and factoring, it's just that breaking up the powershell script into different yaml templates isn't ideal. If you need some organization you can break up the powershell code into functions instead, but still keep it all in one script.

Also, some amount of templates are fine since you don't want to duplicate the logic in each job, so that's why I suggested you just merge `base-image-setup.yaml`, `base-image-build-and-publish.yaml`, and `base-image-remove-librocksdb.yaml` into one yaml template with one powershell script.

Another option is to put the powershell into a ps1 file (which means you could also run it on dev desktop if needed), then call the ps1 script from each job. Either is fine with me.

Comment (yophilav) [reply builds/misc/templates/base-image-publish/base-image-setup.yaml]: Alright, I merge the templates into a template since it's under 100... it should be ok? 
I'm not sure how I feel about having *.ps1 scripts with each executing only a sliver of a for-loop.

File: builds/misc/templates/base-image-publish/base-image-build-and-publish.yaml
Comment (damonbarry) [inline builds/misc/templates/base-image-publish/base-image-build-and-publish.yaml:23]: Nit: There are a couple places in comments where you still refer to the deleted yaml files...

File: builds/misc/templates/base-image-publish/base-image-build-and-publish.yaml
Comment (damonbarry) [inline builds/misc/templates/base-image-publish/base-image-build-and-publish.yaml:36]: `ReturningVstsVar` is no longer a thing...

**Review by damonbarry — DISMISSED** [review]
👍👍

**Review by damonbarry — APPROVED** [review]

**Pull request merged**
